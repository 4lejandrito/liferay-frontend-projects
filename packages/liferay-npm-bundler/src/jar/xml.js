import {js2xml, xml2js} from 'xml-js';

/**
 * Patch a parsed metatype XML file to add pids, localization, etc.
 * @param {string} xml
 * @param {string} localization path of localization file
 * @param {string} pid configuration ID
 * @return {string} the patched XML
 */
export function patchMetatypeXml(xml, {localization, pid}) {
	const js = xml2js(xml, {});

	const metadata = findChild(js, 'metatype:MetaData');

	if (localization) {
		metadata.attributes['localization'] = localization;
	}

	const ocd = findChild(metadata, 'OCD');

	addAttr(ocd, 'id', pid);

	const designate = findChild(metadata, 'Designate', true);

	addAttr(designate, 'pid', pid);

	const object = findChild(designate, 'Object', true);

	addAttr(object, 'ocdref', pid);

	return js2xml(js, {spaces: 2});
}

/**
 * Add an attribute to an XML node.
 * @param {object} node
 * @param {string} name
 * @param {string} value
 */
function addAttr(node, name, value) {
	node.attributes = node.attributes || {};
	node.attributes[name] = value;
}

/**
 * Find an XML child node creating it if necessary.
 * @param {object} parentNode
 * @param {string} childName
 * @param {boolean} create
 * @return {object} the child node
 */
function findChild(parentNode, childName, create = false) {
	let childNode = parentNode.elements.find(node => node.name == childName);

	if (childNode === undefined && create) {
		childNode = {
			type: 'element',
			name: childName,
			attributes: {},
			elements: [],
		};

		parentNode.elements.push(childNode);
	}

	return childNode;
}
