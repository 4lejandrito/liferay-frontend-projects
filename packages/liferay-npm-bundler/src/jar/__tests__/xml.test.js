import fs from 'fs';
import path from 'path';

import {patchMetatypeXml} from '../xml';

describe('patchMetatypeXml', () => {
	it('works with localization file', () => {
		const xmlFile = path.join(__dirname, '__fixtures__', 'metatype.xml');

		const xml = patchMetatypeXml(fs.readFileSync(xmlFile), {
			localization: 'content/Language',
			pid: 'my-portlet',
		});

		expect(xml).toMatchSnapshot();
	});

	it('works without localization file', () => {
		const xmlFile = path.join(__dirname, '__fixtures__', 'metatype.xml');

		const xml = patchMetatypeXml(fs.readFileSync(xmlFile), {
			pid: 'my-portlet',
		});

		expect(xml).toMatchSnapshot();
	});
});
