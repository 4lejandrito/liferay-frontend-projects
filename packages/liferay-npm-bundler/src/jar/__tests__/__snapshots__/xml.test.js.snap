// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`patchMetatypeXml works with localization file 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<metatype:MetaData xmlns:metatype=\\"http://www.osgi.org/xmlns/metatype/v1.1.0\\" localization=\\"content/Language\\">
  <OCD name=\\"My portlet\\" id=\\"my-portlet\\">
    <AD name=\\"a-number\\" id=\\"a-number\\" cardinality=\\"0\\" required=\\"false\\" default=\\"1\\" type=\\"Integer\\" description=\\"a-number-help\\"/>
    <AD name=\\"an-option\\" id=\\"an-option\\" cardinality=\\"0\\" required=\\"false\\" default=\\"A\\" type=\\"String\\" description=\\"an-option-help\\">
      <Option label=\\"option-a\\" value=\\"A\\"/>
      <Option label=\\"option-b\\" value=\\"B\\"/>
    </AD>
  </OCD>
  <Designate pid=\\"my-portlet\\">
    <Object ocdref=\\"my-portlet\\"/>
  </Designate>
</metatype:MetaData>"
`;

exports[`patchMetatypeXml works without localization file 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<metatype:MetaData xmlns:metatype=\\"http://www.osgi.org/xmlns/metatype/v1.1.0\\">
  <OCD name=\\"My portlet\\" id=\\"my-portlet\\">
    <AD name=\\"a-number\\" id=\\"a-number\\" cardinality=\\"0\\" required=\\"false\\" default=\\"1\\" type=\\"Integer\\" description=\\"a-number-help\\"/>
    <AD name=\\"an-option\\" id=\\"an-option\\" cardinality=\\"0\\" required=\\"false\\" default=\\"A\\" type=\\"String\\" description=\\"an-option-help\\">
      <Option label=\\"option-a\\" value=\\"A\\"/>
      <Option label=\\"option-b\\" value=\\"B\\"/>
    </AD>
  </OCD>
  <Designate pid=\\"my-portlet\\">
    <Object ocdref=\\"my-portlet\\"/>
  </Designate>
</metatype:MetaData>"
`;
